lib/semantics/semantics.cc

```diff
@@ -228,6 +230,45 @@ void DoDumpSymbols(std::ostream &os, const Scope &scope, int indent) {
   --indent;
 }
 
+void Semantics::DumpSymbolsSources(std::ostream &os) {
+  std::list<Symbol> symbols;
+  auto less = [&](Symbol a, Symbol b) {
+    return a.name().begin() < b.name().begin();
+  };
+  std::set<Symbol, decltype(less)> visited(less);
```

Comment:
`std::set<const Symbol *>` is probably better. You shouldn't need to copy symbols into the set. Then you probably won't need `less` as well.

Reply:
For the purpose of not printing duplicate definitions, I can't use `Symbol *` because the duplicates aren't the same instances of each other, they just have the same SourceName. I'm not sure of a way to compare their SourceName without using a custom comparator. Maybe I can add an operator < for Symbol but I don't think that really makes sense.

Reply:
Can you give an example where there are two different `Symbol *` that have the same `SourceName`? I'm not sure that should ever happen.

Reply:
For example, if I print out each symbol's address and information in getdefinition01.f90, and then each symbol's source position (without excluding duplicates), I get:
```
0x7fffbd3d1760
m: Module
0x7fffbd3d1d90
f, PRIVATE, PURE, RECURSIVE (Function): Subprogram () result(REAL(4) x)
0x7fffbd3d1970
m, PUBLIC: Misc ScopeName
0x7fffbd3d1b80
s, BIND(C), PUBLIC, PURE (Subroutine): Subprogram (REAL(4) x, REAL(4) yyy)
0x7fffbd3d1fa0
s (Subroutine): HostAssoc
0x7fffbd3d21b0
ss, PURE (Subroutine): Subprogram ()
0x7fffbd3d23c0
x, INTENT(IN) (Implicit): ObjectEntity dummy type: REAL(4)
0x7fffbd3d25d0
yyy, INTENT(INOUT) (Implicit): ObjectEntity dummy type: REAL(4)
0x7fffbd3d27e0
ss (Subroutine): HostAssoc
0x7fffbd3d29f0
f (Function): HostAssoc
0x7fffbd3d2c00
x, ALLOCATABLE: ObjectEntity funcResult type: REAL(4)

m: 18,8,18,9
m: 18,8,18,9
s: 25,18,25,19
f: 37,26,37,27
s: 25,18,25,19
ss: 32,19,32,21
x: 25,21,25,22
yyy: 25,24,25,27
ss: 32,19,32,21
f: 37,26,37,27
x: 39,24,39,25
```

With excluding duplicates using visited, the source position list looks like: 
```
m: 18,8,18,9
f: 37,26,37,27
s: 25,18,25,19
ss: 32,19,32,21
x: 25,21,25,22
yyy: 25,24,25,27
x: 39,24,39,25
```

It seems like all of these symbols found are different from each other, but they can resolve into the same SourceName.

Reply:
Change DumpSymbolsSources to use a map between Symbol's SourceName and pointer, thanks to Tim's suggestions and patch.